Add chat templates integration

juliendenize opened the pull request:
This PR aims to add chat templates inside this repo as a valid working implementation in comparison with mistral-common to work with libraries that require jinja templates (llama.cpp, vanilla Transformers, SG Lang ...).

This is in order to give the possibility to our community to use our models beyond the scope of mistral-common when the latter is not integrated in parts of the ecosystem.

We validate these chat templates on a set of dummy requests.

Next step is to add tests for edge cases and errors + integration tests on some datasets.

joa-stdn commented on tests/data/chat_templates/v15_image.jinja:
```diff
@@ -0,0 +1,204 @@
+        {%- elif message['content'] | length > 0 %}
+            {{- '[INST]' }}
+            {%- if message['content'] | length == 2 %}
+                {%- set blocks = message['content'] | sort(attribute='type') %}
```
is this because of image reordering? written this way is weird no?

joa-stdn commented on tests/data/chat_templates/v15_image.jinja:
```diff
@@ -0,0 +1,204 @@
+                {{- raise_exception('Unexpected role \'' + current_role + '\' after role \'' + ns_order.previous_role + '\'') }}
+            {%- endif %}
+        {%- elif ns_order.previous_role == 'user' %}
+            {%- if current_role != 'assistant' and current_role != 'system' and current_role != 'user' %}
```
maybe better to do equality here?

joa-stdn commented on tests/data/chat_templates/v15_image.jinja:
```diff
@@ -0,0 +1,204 @@
+                            {%- set ns_c.text_parts = ns_c.text_parts + [block['text']] %}
+                        {%- else %}
+                            {%- if ns_c.text_parts | length > 0 %}
+                                {%- set ns_c.chunks = ns_c.chunks + [{'type': 'text', 'text': ns_c.text_parts | join('\n\n')}] %}
```
actually internally we only aggregate across messages now but used indeed to aggregate for chunks in the same message

juliendenize replied:
should be something that i change in the end but should be first deprecated

joa-stdn commented on tests/data/chat_templates/v15_image.jinja:
```diff
@@ -0,0 +1,204 @@
+                {%- if block['type'] == 'text' %}
+                    {{- block['text'] }}
+                {%- else %}
+                    {{- raise_exception('Only text chunks are supported in system message contents.') }}
```
do you think we could have a macro for some of those things? like render_content?

joa-stdn commented on tests/data/chat_templates/v15_image_think.jinja:
```diff
@@ -0,0 +1,224 @@
+                    {{- block['text'] }}
+                {%- elif block['type'] == 'thinking' %}
+                    {{- '[THINK]' + block['thinking'] }}
+                    {%- if block.get('closed', true) %}{{- '[/THINK]' }}{%- endif %}
```
tbh it should really always be true IMO, are we sure it's going to be ok?

juliendenize replied:
it's an option from mistral-common so i'd like to replicate, think it should be fine anyway as i think very few people would pass this flag.

juliendenize committed:
Add HuggingFace chat template generation and integration tests

joa-stdn commented on .github/workflows/lint_build_test.yaml:
```diff
@@ -95,3 +95,32 @@ jobs:
+          "3.14",
+        ]
+    steps:
+      - name: Checkout
```
are we ok with github actions?

juliendenize replied:
i should move to bk probably eventually. will adress in followup

joa-stdn commented on src/mistral_common/integrations/chat_templates/chat_templates.py:
```diff
@@ -0,0 +1,60 @@
+    image_support: bool,
+    audio_support: bool,
+    thinking_support: bool,
+    default_system_prompt: str | None = None,
```
let's remove defaults?

joa-stdn commented on src/mistral_common/integrations/chat_templates/chat_templates.py:
```diff
@@ -0,0 +1,60 @@
+    template = _build_chat_template(config)
+
+    if default_system_prompt is not None:
+        escaped_prompt = default_system_prompt.replace("\\", "\\\\").replace("'", "\\'")
```
we're not afraid of the thinking chunk here? Worth a comment at least maybe?

juliendenize replied:
wdym ? think chunks doesn't need escape right ?
Or do you mean that users should keep in mind to add think tokens if needed ?

joa-stdn replied:
I meant system prompt themselves can contain chunks? But here it's str, should be ok

joa-stdn commented on src/mistral_common/integrations/chat_templates/template_generator.py:
```diff
@@ -0,0 +1,1674 @@
+    audio_support: bool = False
+    thinking_support: bool = False
+    plain_thinking_support: bool = False
+    use_token_variables: bool = True
```
this name is very weird, I wonder if we can maybe find a better name no?

juliendenize replied:
changed the name lmk if it looks better :)

joa-stdn commented on src/mistral_common/integrations/chat_templates/template_generator.py:
```diff
@@ -0,0 +1,1674 @@
+        r"""Jinja expression for the BOS token."""
+        if self.use_token_variables:
+            return "bos_token"
+        return "'" + _BOS + "'"
```
is this better than f strings?

juliendenize replied:
nop nice catch

joa-stdn commented on src/mistral_common/integrations/chat_templates/template_generator.py:
```diff
@@ -0,0 +1,1674 @@
+    else:
+        # pre-v7: system messages are already filtered out
+        if config.has_tools:
+            if config.version == TokenizerVersion.v2 or (config.spm and config.version == TokenizerVersion.v3):
```
to me it's quite weird to have both explicit versions and the simplifying properties, we should go one way or the other no?

juliendenize replied:
true, added more properties

joa-stdn commented on tests/integrations/chat_templates/transformers/test_core_parity.py:
```diff
@@ -0,0 +1,310 @@
+                        assert len(message.content) == 1 and isinstance(message.content[0], TextChunk), (
+                            "Only text content is supported for v1 and v2"
+                        )
+                        message.content = str(message.content[0].text)
```
is it better to do this or to have them already in str?

juliendenize replied:
better like this as i also want to test chunks

joa-stdn commented on tests/integrations/chat_templates/transformers/test_core_parity.py:
```diff
@@ -0,0 +1,310 @@
+            if version == TokenizerVersion.v2:
+                for message in conversation.messages:
+                    if isinstance(message, ToolMessage):
+                        message.name = "tool"
```
I think we can just keep names all the time?

juliendenize replied:
true

joa-stdn commented on tests/integrations/chat_templates/transformers/test_core_parity.py:
```diff
@@ -0,0 +1,310 @@
+            ]
+        }
+        # This should not raise
+        encode_transformers(chat_template, valid_consecutive_users)
```
why don't we check matching here?

juliendenize replied:
these tests are redundant because already tested valid ones in the conversations, going to remove.

joa-stdn commented on tests/integrations/chat_templates/transformers/test_core_parity.py:
```diff
@@ -0,0 +1,310 @@
+        think: bool,
+    ) -> None:
+        invalid_sp_think = {
+            "messages": [
```
maybe let's define this in fixtures / outside of tests and have the test itself shorter?

joa-stdn commented on tests/integrations/chat_templates/transformers/test_core_parity.py:
```diff
@@ -0,0 +1,310 @@
+    ) -> None:
+        conversations = _get_conversations(version, mode, image, audio, think)
+
+        mistral_tokenizer = _get_mistral_tokenizer(
```
curious why we don't do the MistralTokenizer.from_file directly with a file in tmp_path etc?

juliendenize replied:
we don't have all the versions in the fixtures data. I could download some official but decided instead to build them on runtime. Is it a blocker to you ?

juliendenize replied:
in the end i changed it to be able to use MistralCommonBackend :)

joa-stdn commented on tests/integrations/chat_templates/conftest.py:
```diff
@@ -0,0 +1,86 @@
+
+# All configurations for output comparison tests (including SPM).
+# Tuples: (version, spm, image, audio, think, plain_think)
+ALL_CONFIGS: list[tuple[TokenizerVersion, bool, bool, bool, bool, bool]] = [
```
should we dataclass for this?

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,340 @@
+    r"""Encode a chat request using mistral-common tokenizer."""
+    mistral_encoded = str(mistral_tokenizer.encode_chat_completion(chat_request).text)
+    # Remove image tokens except one per image
+    mistral_encoded = mistral_encoded.replace("[IMG]", "").replace("[IMG_BREAK]", "").replace("[IMG_END]", "[IMG]")
```
Actually doesn't it make more sense to concatenate all [IMG] into [IMG] and drop the two others? Maybe we don't care, this is nit

juliendenize replied:
harder to setup if you have multiple images that are one after the other

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,340 @@
+
+def encode_mistral_common(mistral_tokenizer: MistralTokenizer, chat_request: ChatCompletionRequest, spm: bool) -> str:
+    r"""Encode a chat request using mistral-common tokenizer."""
+    mistral_encoded = str(mistral_tokenizer.encode_chat_completion(chat_request).text)
```
there's no way to compare token ids? I thought this was the purpose...

juliendenize replied:
it is possible but requires to use apply_chat_template from HF. I could add this indeed.

juliendenize replied:
added

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,340 @@
+    }
+
+    # Only add reasoning_effort for v15+ templates that support it
+    if reasoning_effort is not None or "reasoning_effort" in template:
```
seems weird to me, I would have said that we need to restructure the tests so that this line can just be
```
if reasoning_effort is not None
```
no?

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,340 @@
+
+
+def encode_transformers(
+    chat_template: str, chat_request: ChatCompletionRequest | dict[str, Any], keep_name_for_tools: bool = False
```
why do we allow both pydantic object and ChatCompletionRequest

juliendenize replied:
the other one was to test invalid or dict request from openai, created another function

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,340 @@
+    if reasoning_effort is not None:
+        template_kwargs["reasoning_effort"] = reasoning_effort
+
+    encoded = render_jinja_template(
```
maybe add a comment on why we have render_template as well?

juliendenize replied:
there is a docstring for the function that should explain it

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,340 @@
+            raise ValueError(f"Unknown tokenizer version: {tokenizer_version}")
+
+
+def _get_mistral_tekkenizer(
```
why don't we use MistralTokenizer.from_file?

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,340 @@
+    vocab = json_tekkenizer["vocab"]
+    vocab_size = json_tekkenizer["config"]["default_vocab_size"]
+    pattern = json_tekkenizer["config"]["pattern"]
+    model_settings_builder = (
```
this is too hacky no?

juliendenize replied:
why do you think that ?

juliendenize replied:
Refactored — now uses `MistralTokenizer.from_file` via `_build_tekken_json` + `_get_mistral_tekkenizer`. No more manual tokenizer construction.

juliendenize committed:
refactor: rename use_token_variables to use_special_token_variables and remove defaults from generate_chat_template

juliendenize committed:
refactor: use f-strings for BOS/EOS token expressions

juliendenize committed:
refactor: replace direct version checks with named properties in TemplateConfig

juliendenize committed:
docs: add explanatory comments addressing review feedback

juliendenize committed:
refactor: split encode_transformers, simplify reasoning_effort check

juliendenize committed:
refactor: always set tool names, add consecutive users parity check, extract test fixtures

juliendenize committed:
refactor: replace config tuples with TestConfig dataclass

juliendenize committed:
refactor: remove redundant valid-case checks from test_role_error

juliendenize committed:
refactor: convert INVALID_* constants to pytest fixtures

juliendenize committed:
feat: add token ID parity comparison via MistralCommonBackend

juliendenize committed:
fix: reorder ThinkChunks to leading position and update doctest

juliendenize committed:
refactor: replace MistralCommonBackend with PreTrainedTokenizerFast for token parity

Use a standard HuggingFace PreTrainedTokenizerFast built from tekken.json
instead of MistralCommonBackend for token ID comparison. This provides an
independent verification using the Rust tokenizers library rather than a
circular test through mistral-common.

Skip token ID comparison for image, audio, and SPM configs.

juliendenize committed:
fix: add tokenizers to mypy ignore_missing_imports

joa-stdn commented on src/mistral_common/integrations/chat_templates/template_generator.py:
```diff
@@ -0,0 +1,1698 @@
+    Aggregates consecutive messages with the same role before the alternation
+    check to match the behavior of the mistral-common normalizer:
+
+    - **User messages**: consecutive messages merged into one, text content joined
```
we need to nuke this actually, any chance we can do that in follow up?

joa-stdn commented on src/mistral_common/integrations/chat_templates/template_generator.py:
```diff
@@ -0,0 +1,1698 @@
+    - **User messages**: consecutive messages merged into one, text content joined
+      with `"\\n\\n"`, non-text chunks (image, audio) preserved in order.
+    - **Assistant messages**: consecutive messages merged into one, text content joined
+      with `"\\n\\n"`, `tool_calls` lists concatenated.
```
all of this is no longer up to date in new models I think

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,695 @@
+        audio_encoder=audio_encoder,
+    )
+    model_settings_builder = tokenizer.model_settings_builder if isinstance(tokenizer, Tekkenizer) else None
+    return MistralTokenizer(
```
I think MistralTokenizer.from_file is more natural, we have that in test_tokenizer_v14 etc no?

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,695 @@
+        image_encoder=image_encoder,
+        audio_encoder=audio_encoder,
+    )
+    model_settings_builder = tokenizer.model_settings_builder if isinstance(tokenizer, Tekkenizer) else None
```
this would allow not doing this

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,695 @@
+    return active_ids
+
+
+def _build_hf_tokenizer(
```
can do we a file for hf_utils maybe? This one is quite big?

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,695 @@
+    if num_special == 0:
+        return active_ids
+
+    for attr_name in vars(instruct_tokenizer):
```
not super clean no? why don't we just add all special tokens to the HF model? this is what we do anyways in Tekkenizer?

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,695 @@
+    openai_request = chat_request.to_openai()
+    if keep_name_for_tools:
+        for openai_message, chat_message in zip(openai_request["messages"], chat_request.messages):
+            if chat_message.role == "tool":
```
this is duplicated can we factorize?

juliendenize committed:
refactor: extract _to_openai_request to factorize keep_name_for_tools logic

juliendenize committed:
refactor: register all special tokens in HF tokenizer, remove introspection

juliendenize committed:
refactor: use MistralTokenizer.from_file in test helpers

juliendenize committed:
refactor: split HF-specific helpers into hf_utils.py

joa-stdn commented on tests/integrations/chat_templates/helpers.py:
```diff
@@ -0,0 +1,384 @@
+        tokenizer=tokenizer,
+        image_encoder=image_encoder,
+    )
+    return MistralTokenizer(
```
why not from file here?

juliendenize committed:
refactor: use MistralTokenizer.from_file for SPM backend in test helpers
